File: js/plain/p4.js
'use strict'

function ReverseString(s)
{
    var rev = ''

    for (var i = s.length - 1; i >= 0; i--)
    {
        rev += s[i]
    }

    return rev
}

function IsStringPalindrome(s)
{
    var left = 0
    var right = s.length - 1

    while (left < right)
    {
        if (s[left] != s[right])
        {
            return false
        }

        left++
        right--
    }

    return true
}

function IsPalindrome(n)
{
    return IsStringPalindrome(n + '')
}

var largest = 0

for (var i = 100; i <= 999; i++)
{
    for (var j = i; j <= 999; j++)
    {
        var ij = i * j
        if (IsPalindrome(ij) && ij > largest)
        {
            largest = ij
        }
    }
}

console.log(largest)

File: cpp/p4.cpp
#include <deque>
#include <iostream>
using namespace std;

bool IsPalindrome(int x)
{
    deque<int> digits;
    while (x != 0)
    {
        digits.push_back(x % 10);
        x /= 10;
    }
    
    int left = 0;
    int right = digits.size() - 1;
    while (left < right)
    {
        if (digits[left] != digits[right])
            return false;
        
        ++left;
        --right;
    }
    
    return true;
}

int main()
{
    int largest = -1;
    for (int i = 100; i != 1000; ++i)
        for (int j = 100; j != 1000; ++j)
            if (IsPalindrome(i * j) && i * j > largest)
                largest = i * j;
    
    cout << largest << endl;
    
    return 0;
}
